### DISCUSSION TITLE
Nested tree-based navigation issue

### DISCUSSION THREAD

**Discussion by teamboot88** [pointfreeco/swift-composable-architecture#3007] [Q&A] [answered]
Hello, 
I am a SwiftUI newbie who is diligently and wholeheartedly using TCA. 
First of all, thank you very much.

Unfortunately, I’m facing some issues when implementing nested tree-based navigation with a structure like 
`PhotoList` -> `PhotoPicker` -> `PhotoCropper`  

Specifically, I’m encountering a warning when attempting to dismiss PhotoPicker and PhotoCropper simultaneously and present PhotoForm when the “cropCompleted” action occurs in PhotoCropper. 
`PhotoList` -> `PhotoForm`

(To detect the “cropCompleted” action in PhotoCropper from PhotoList, I am doing the following.)
```swift
// Unnecessary code was elminated.

import ComposableArchitecture
import Foundation

extension PhotoListFeature {
    @Reducer(state: .equatable)
    enum Destination {
        case photoPicker(PhotoPickerFeature)
        case photoForm(PhotoFormFeature)
    }
}

@Reducer
struct PhotoListFeature {
    @ObservableState
    struct State: Equatable {
        @Presents
        var destination: Destination.State?
    }

    enum Action {
        case destination(PresentationAction<Destination.Action>)
    }

    var body: some ReducerOf<Self> {
        Reduce { state, action in
            switch action {
            case let .destination(.presented(.photoPicker(.destination(.presented(.photoCropper(.cropCompleted(uiImage: uiImage))))))):

                // here !
                state.destination = .photoForm(
                    PhotoFormFeature.State(uiImage: uiImage)
                )

                return .none

            case .destination:
                return .none
            }
        }
        .ifLet(\.$destination, action: \.destination)
    }
}
```

The problem is that when replacing the destination as described above, the following warning occurs.
```
"Treest/PhotoListFeature.swift" received a child action when child state was set to a different case. …
  Action:
    PhotoListFeature.Destination.Action.photoPicker(.destination(.dismiss))
  State:
    PhotoListFeature.Destination.State.photoForm
```

In this nested form of navigation, when replacing the destination from an ancestor (more than two levels deep, not the parent), it is suspected that the event of dismissing the destination of the child, PhotoPicker, is triggered, resulting in the above warning.

- Is it safe to ignore this warning? 
- If you have any advice on the correct approach for dismissing multiple children at once or replacing them with another feature in nested tree-based navigation, I would greatly appreciate it. _ _)

**Comment by mbrandonw** [comment, marked as answer]
> In this nested form of navigation, when replacing the destination from an ancestor (more than two levels deep, not the parent), it is suspected that the event of dismissing the destination of the child, PhotoPicker, is triggered, resulting in the above warning.

Hi @teamboot88, this is happening due to bugs in SwiftUI navigation. When you dismiss two presented features at once, the intermediate view incorrectly writes `nil` to its binding, causing a `.dismiss` action to be sent when it should not.

> Is it safe to ignore this warning?

Yeah, for this particular situation it is ok to ignore because it is sadly just a SwiftUI bug.

> If you have any advice on the correct approach for dismissing multiple children at once or replacing them with another feature in nested tree-based navigation, I would greatly appreciate it. _ _)

What you have here should be fine, but there are two potential workarounds:

* Stagger the dismissal so that you dismiss the innermost feature, wait a small amount of time (like 0.5 seconds) and then dismiss the outer feature.
* [Stack-based navigation](https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/stackbasednavigation) tends to have fewer bugs in SwiftUI, and so that is also an option.

**Reply by teamboot88** [reply]
Oh, that's a bug in SwiftUI. I'll refer to the workaround. Thank you so much for your quick response.
